pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -632,9 +692,48 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 		changes.KillPod = true
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarLifecycle) {
+		if pod.Spec.RestartPolicy != v1.RestartPolicyAlways {
+			//Determine if there are any non sidecar containers that are running or need restarting
+			//if there are none, we can kill the remaining sidecars
+			onlySidecars := onlySidecars(pod, podStatus)
+
+			//only sidecars are left so terminate them all
+			if onlySidecars {
+				changes.KillPod = true
```

Comment:
If I'm reading `SyncPod` correctly, I think that if `changes.KillPod == true`, then it does not matter what containers are included in `ContainersToKill` - as all containers are killed via the `killPodWithSyncResult` method. So I'm a little confused about why we are adding the sidecar containers to `ContainersToKill`? Is it just to handle future behavior changes from `SyncPod`?

Reply:
If I read the PR correctly, kubeGenericRuntimeManager#killContainersWithSyncResult distinguishes sidecars versus nonSidecars. The sidecars would be killed after nonSidecars.

Reply:
Yeah you're right @mattjmcnaughton, the containers to kill is not actually used due to kill pod being true, I think it's actually nicer if we do use `ContainersToKill` as it means you get a message on your pod saying "All containers have permanently exited, sidecar container %s will be killed".  
The kill pod was there more as a catch to make sure we kill the pod if another sync happens and all the sidecars have shut down.
So could change this code to evaluate at the end of this sidecars loop with 
```go
if len(changes.ContainersToKill) == 0 {
  changes.KillPod == true
}
```